tools/thanosconvert/thanosconvert.go

```diff
@@ -0,0 +1,188 @@
+	level.Info(c.logger).Log("msg", "Scanned tenants")
+
+	for _, u := range users {
+		r, err := c.convertUser(ctx, u)
+		results[u] = r
+		if err != nil {
+			return results, errors.Wrap(err, fmt.Sprintf("error converting user %s", u))
+		}
+	}
+	return results, nil
+}
+
+func (c ThanosBlockConverter) convertUser(ctx context.Context, user string) (PerUserResults, error) {
+	results := PerUserResults{}
+
+	userBucketClient := bucket.NewUserBucketClient(user, c.bkt)
+	err := userBucketClient.Iter(ctx, "", func(o string) error {
+		// get block ID from path
+		o = strings.TrimSuffix(o, "/")
+		blockULID, err := ulid.Parse(o)
+		if err != nil {
+			// this is not a block, skip it
+			return nil
+		}
+		blockID := blockULID.String()
+
+		level.Debug(c.logger).Log("msg", "Processing block", "block", blockID)
+
+		// retrieve meta.json
+
+		metaKey := fmt.Sprintf("%s/%s/meta.json", user, blockID)
+		r, err := c.bkt.Get(ctx, metaKey)
+		if err != nil {
+			level.Error(c.logger).Log("msg", "Get meta.json", "block", blockID, "meta_key", metaKey, "err", err.Error())
+			results.AddFailed(blockID)
+			return nil
+		}
+
+		meta, err := metadata.Read(r)
+		if err != nil {
+			level.Error(c.logger).Log("msg", "decode meta.json into metadata.Meta", "block", blockID, "meta_key", metaKey, "err", err.Error())
+			results.AddFailed(blockID)
+			return nil
+		}
+
+		// convert and upload if appropriate
+
+		newMeta, changesRequired := convertMetadata(*meta, user)
+
+		if len(changesRequired) > 0 {
+			if c.dryRun {
+				level.Debug(c.logger).Log("msg", "Block requires changes, not uploading due to dry run", "block", blockID, "changes_required", strings.Join(changesRequired, ","))
+			} else {
+				level.Debug(c.logger).Log("msg", "Block requires changes, uploading new meta.json", "block", blockID, "changes_required", strings.Join(changesRequired, ","))
+				if err := c.uploadNewMeta(ctx, userBucketClient, blockID, newMeta); err != nil {
+					level.Error(c.logger).Log("msg", "Update meta.json", "block", blockID, "meta_key", metaKey, "err", err.Error())
+					results.AddFailed(blockID)
+					return nil
+				}
+			}
+			results.AddConverted(blockID)
+		} else {
+			results.AddUnchanged(blockID)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return results, err
+	}
+
+	return results, nil
+}
+
+func (c *ThanosBlockConverter) uploadNewMeta(ctx context.Context, userBucketClient *bucket.UserBucketClient, blockID string, meta metadata.Meta) error {
+	var body bytes.Buffer
+	if err := meta.Write(&body); err != nil {
+		return errors.Wrap(err, "encode json")
+	}
+
+	if err := userBucketClient.Upload(ctx, fmt.Sprintf("%s/meta.json", blockID), &body); err != nil {
+		return errors.Wrap(err, "upload to bucket")
+	}
+	return nil
+}
+
+func convertMetadata(meta metadata.Meta, expectedUser string) (metadata.Meta, []string) {
+	var changesRequired []string
+
+	if meta.Thanos.Labels == nil {
+		// TODO: no "thanos" entry could mean this block is corrupt - should we bail out instead?
```

Comment:
I would not bail out. If we don't require the Thanos entry, the tool should work with Prometheus blocks too, which I believe is a plus for the community.